docs/root/configuration/http/http_conn_man/headers.rst

```diff
@@ -574,271 +574,18 @@ used to delimit variable names.
-
-%HOSTNAME%
-    The system hostname.
-
-%PROTOCOL%
-    The original protocol which is already added by Envoy as a
-    :ref:`x-forwarded-proto <config_http_conn_man_headers_x-forwarded-proto>` request header.
-
-%REQUESTED_SERVER_NAME%
-  HTTP
-    String value set on ssl connection socket for Server Name Indication (SNI)
-  TCP
-    String value set on ssl connection socket for Server Name Indication (SNI)
-
-%UPSTREAM_METADATA(["namespace", "key", ...])%
-    Populates the header with :ref:`EDS endpoint metadata <envoy_v3_api_field_config.endpoint.v3.LbEndpoint.metadata>` from the
-    upstream host selected by the router. Metadata may be selected from any namespace. In general,
-    metadata values may be strings, numbers, booleans, lists, nested structures, or null. Upstream
-    metadata values may be selected from nested structs by specifying multiple keys. Otherwise,
-    only string, boolean, and numeric values are supported. If the namespace or key(s) are not
-    found, or if the selected value is not a supported type, then no header is emitted. The
-    namespace and key(s) are specified as a JSON array of strings. Finally, percent symbols in the
-    parameters **do not** need to be escaped by doubling them.
-
-    Upstream metadata cannot be added to request headers as the upstream host has not been selected
-    when custom request headers are generated.
-
-%DYNAMIC_METADATA(["namespace", "key", ...])%
-    Similar to UPSTREAM_METADATA, populates the header with dynamic metadata available in a request
-    (e.g.: added by filters like the header-to-metadata filter).
-
-    This works both on request and response headers.
-
-%UPSTREAM_LOCAL_ADDRESS%
-    Local address of the upstream connection. If the address is an IP address it includes both
-    address and port.
-
-    The upstream local address cannot be added to request headers as the upstream host
-    hremote as not been selected when custom request headers are generated.
-
-%UPSTREAM_LOCAL_ADDRESS_WITHOUT_PORT%
-    Local address of the upstream connection, without any port component.
-    IP addresses are the only address type with a port component.
-
-%UPSTREAM_LOCAL_PORT%
-    Local port of the upstream connection.
-    IP addresses are the only address type with a port component.
-
-%UPSTREAM_REMOTE_ADDRESS%
-    Remote address of the upstream connection. If the address is an IP address it includes both
-    address and port.
-
-    The upstream remote address cannot be added to request headers as the upstream host
-    has not been selected when custom request headers are generated.
-
-%UPSTREAM_REMOTE_ADDRESS_WITHOUT_PORT%
-    Remote address of the upstream connection, without any port component.
-    IP addresses are the only address type with a port component.
-
-%UPSTREAM_REMOTE_PORT%
-    Remote port of the upstream connection.
-    IP addresses are the only address type with a port component.
-
-%PER_REQUEST_STATE(reverse.dns.data.name)%
```

Comment:
are we losing any information here?

not all of these substitution vars seem to appear on the linked page

Reply:
I am using term `header formatters` to describe the separate formatters which were documented in docs/root/configuration/http/http_conn_man/headers.rst. 
As `access log formatters` I mean formatters previously used to create access log entry. 
The idea behind this PR is to use` access log formatters` in both scenarios.
The situation is as follows: all formatters which were separately documented as `header formatters` are also present in access log formatters in exact form. There are 3 exceptions: `header formatter's` DYNAMIC_METADATA and UPSTREAM_METADATA used to take the same parameters as access log equivalent, but in different form.
`header formatter's` PER_REQUEST_STATE is not part of access log formatters (as you noted above), but the same information is accessed using access log's FILTER_STATE formatter.
In order to assure backwards compatibility I added code to translate DYNAMIC_METADATA and UPSTREAM_METADATA from "old" header formatter's form to the one used by access log. Both forms can be used now. Also, I added a translator to translate header formatter's PER_REQUEST_STATE to FILTER_STATE. As translator I mean C++ Envoy code which kicks in when config is received, not an offline translator which processes config files.
It means that whoever is using header formatter's format of *_METADATA and PER_REQUEST_STATE can continue doing so, until it is decided to deprecate those formats.
But I think that from now on, all header formatters should use formats equivalent to access log's one.

Do you think I should document all that logic in docs or release notes?

Reply:
probably we need to say something in the changelog, but im still concerned about just removing these from the docs

i guess my first question (hoping iuc above) - are we deprecating the older substitution vars now ? (or have we already)

i think ideally we want something in the docs themselves showing the vars as deprecated for a period of time

Reply:
i understand about the de-duplication here - i think defo a good idea

my concern is that either 

a) we are keeping `PER_REQUEST_STATE` + others, but we wont have any docs for them, or
b) we are not keeping them, but we are removing them from the docs before deprecating them

either way, if someone were to search for `PER_REQUEST_STATE` now they _would_ get hits, whereas after this PR they would not

Reply:
@plax Thanks a lot for your comments and for digging into details of this PR! I agree with you that some sort of documentation is needed about formatters which changed and are going to be deprecated. I added `attention` section listing mentioned 3 formatters and substitutes which should be used.